src/DiagnosticMessages.ts

```diff
@@ -623,15 +626,22 @@ export let DiagnosticMessages = {
     })
 };
 
-let allCodes = [] as number[];
+export const DiagnosticCodeMap = {} as Record<keyof (typeof DiagnosticMessages), number>;
+export let diagnosticCodes = [] as number[];
 for (let key in DiagnosticMessages) {
-    allCodes.push(DiagnosticMessages[key]().code);
+    diagnosticCodes.push(DiagnosticMessages[key]().code);
+    DiagnosticCodeMap[key] = DiagnosticMessages[key]().code;
 }
 
-export let diagnosticCodes = allCodes;
-
 export interface DiagnosticInfo {
     message: string;
     code: number;
     severity: DiagnosticSeverity;
 }
+
+/**
+ * Provides easy type support for the return value of any DiagnosticMessage function.
+ * The second type parameter is optional, but allows plugins to pass in their own
+ * DiagnosticMessages-like object in order to get the same type support
+ */
+export type DiagnosticMessageType<K extends keyof D, D extends Record<string, (...args: any) => any> = typeof DiagnosticMessages> = ReturnType<D[K]>;
```

Comment:
@elsassph I think you'll find this quite useful. It allows you to do this:
```
const diagnostic = someDiagnostic as DiagnosticMessageType<'callToUnknownFunction'>;
```
which will look at DiagnosticMessage.callToUnknownFunction(), and enforce type checking based on that return value. This is useful when paired with the new `BsDiagnostic.data` `any` object, so we can enforce stricter types when using those.

This also supports an optional second type parameter, which allows plugins to use the same functionality, assuming they follow the same DiagnosticMessages model in their own plugins. 
```
export const BsLintDiagnosticMessages = {
    someDiagnostic: ()=>({
        message: 'Some diagnostic',
        code: 12345,
        data: {
            isTrue: true
        }
    })
};

const typedDiagnostic= program.getDiagnostics()[0] as DiagnosticMessageTypes<'someDiagnostic', BsDiagnosticMessages>;

Reply:
You can see it used [here](https://github.com/rokucommunity/brighterscript/pull/298/files#diff-1412e36a3e7918cf699241fb44ab227a45ce0b7d92f609b4d543021861252d93R28)